File: sudoku-game/game-boot/src/main/java/com/zumin/sudoku/game/utils/SudokuSolver.java
package com.zumin.sudoku.game.utils;

import java.util.ArrayList;
import java.util.List;
import lombok.Getter;

/**
 * 数独游戏解决器
 */
public class SudokuSolver {

  private static final int THRESHOLD = 324;
  @Getter
  private final List<int[][]> solutions = new ArrayList<>();
  private final int[] size = new int[1074];
  private final DLXNode[] row = new DLXNode[4146];
  private final DLXNode[] col = new DLXNode[1074];
  private final int[][] data = new int[9][9];
  private int maxDepth;
  private DLXNode head;

  /**
   * 公有构造方法
   *
   * @param topic    数独题目数组
   * @param maxDepth 最大深度
   */
  public SudokuSolver(int[][] topic, int maxDepth) {
    init(topic, maxDepth);
  }

  /**
   * 公有构造方法，默认最大深度为2
   *
   * @param topic 数独题目数组
   */
  public SudokuSolver(int[][] topic) {
    init(topic, 2);
  }

  /**
   * 获取数独题解个数
   *
   * @return 数独题解个数
   */
  public int solutionCount() {
    return solutions.size();
  }

  /**
   * 初始化该类，并解决数独题目
   *
   * @param topic    数独题目
   * @param maxDepth 最大深度
   */
  private void init(int[][] topic, int maxDepth) {
    this.maxDepth = maxDepth;
    initDLX();
    initData(topic);
    solve(0);
  }

  /**
   * 初始化DLX节点
   */
  private void initDLX() {
    int r = 730, c = 324;
    head = new DLXNode(r, c);
    head.U = head.D = head.L = head.R = head;
    for (int i = 0; i < c; ++i) {
      col[i] = new DLXNode(r, i);
      col[i].L = head;
      col[i].R = head.R;
      col[i].L.R = col[i].R.L = col[i];
      col[i].U = col[i].D = col[i];
      size[i] = 0;
    }

    for (int i = r - 1; i > -1; --i) {
      row[i] = new DLXNode(i, c);
      row[i].U = head;
      row[i].D = head.D;
      row[i].U.D = row[i].D.U = row[i];
      row[i].L = row[i].R = row[i];
    }
  }

  /**
   * 添加DLX节点
   *
   * @param r 行
   * @param c 列
   */
  private void addNode(int r, int c) {
    DLXNode p = new DLXNode(r, c);
    p.R = row[r];
    p.L = row[r].L;
    p.L.R = p.R.L = p;
    p.U = col[c];
    p.D = col[c].D;
    p.U.D = p.D.U = p;
    ++size[c];
  }

  private void addNode(int i, int j, int k) {
    int r = (i * 9 + j) * 9 + k;
    addNode(r, i * 9 + k - 1);
    addNode(r, 9 * 9 + j * 9 + k - 1);
    addNode(r, 2 * 9 * 9 + block(i, j) * 9 + k - 1);
    addNode(r, 3 * 9 * 9 + i * 9 + j);
  }

  private int block(int x, int y) {
    return x / 3 * 3 + y / 3;
  }

  private void cover(int c) {
    if (c == THRESHOLD) {
      return;
    }

    col[c].delLR();
    DLXNode R, C;
    for (C = col[c].D; C != col[c]; C = C.D) {
      if (C.c == THRESHOLD) {
        continue;
      }
      for (R = C.L; R != C; R = R.L) {
        if (R.c == THRESHOLD) {
          continue;
        }
        --size[R.c];
        R.delUD();
      }
      C.delLR();
    }
  }

  private void resume(int c) {
    if (c == THRESHOLD) {
      return;
    }

    DLXNode R, C;
    for (C = col[c].U; C != col[c]; C = C.U) {
      if (C.c == THRESHOLD) {
        continue;
      }
      C.resumeLR();
      for (R = C.R; R != C; R = R.R) {
        if (R.c == THRESHOLD) {
          continue;
        }
        ++size[R.c];
        R.resumeUD();
      }
    }
    col[c].resumeLR();
  }

  private boolean solve(int depth) {
    if (head.L == head) {
      int[][] solution = new int[9][9];
      for (int i = 0; i < 9; ++i) {
        System.arraycopy(data[i], 0, solution[i], 0, 9);
      }
      solutions.add(solution);

      return solutions.size() >= this.maxDepth;
    }
    int minSize = 1 << 30;
    int c = -1;
    DLXNode p;
    for (p = head.L; p != head; p = p.L) {
      if (size[p.c] < minSize) {
        minSize = size[p.c];
        c = p.c;
      }
    }
    cover(c);

    for (p = col[c].D; p != col[c]; p = p.D) {
      DLXNode cell;
      p.R.L = p;
      for (cell = p.L; cell != p; cell = cell.L) {
        cover(cell.c);
      }
      p.R.L = p.L;
      int rr = p.r - 1;
      data[rr / (9 * 9)][rr / 9 % 9] = rr % 9 + 1;
      if (solve(depth + 1)) {
        return true;
      }

      p.L.R = p;
      for (cell = p.R; cell != p; cell = cell.R) {
        resume(cell.c);
      }
      p.L.R = p.R;
    }

    resume(c);
    return false;
  }

  private void initData(int[][] topic) {
    int i, j, k;
    for (i = 0; i < 9; ++i) {
      for (j = 0; j < 9; ++j) {
        if (topic[i][j] > 0) {
          addNode(i, j, topic[i][j]);
        } else {
          for (k = 1; k <= 9; ++k) {
            addNode(i, j, k);
          }
        }
      }
    }
  }

  private static class DLXNode {

    private final int r;
    private final int c;
    private DLXNode U, D, L, R;

    public DLXNode(int r, int c) {
      this.r = r;
      this.c = c;
    }

    public void delLR() {
      L.R = R;
      R.L = L;
    }

    public void delUD() {
      U.D = D;
      D.U = U;
    }

    public void resumeLR() {
      L.R = R.L = this;
    }

    public void resumeUD() {
      U.D = D.U = this;
    }
  }
}

File: sudoku-game/game-boot/src/main/java/com/zumin/sudoku/game/utils/SudokuSolver.kt
package com.zumin.sudoku.game.utils

private const val THRESHOLD = 324

/**
 * 数独游戏解决器
 */
class SudokuSolver(
  topic: Array<IntArray>,
  private val maxDepth: Int = 2,
) {
  private val solutions: MutableList<Array<IntArray>> = ArrayList()
  private val size = IntArray(1074)
  private val row = arrayOfNulls<DLXNode>(4146)
  private val col = arrayOfNulls<DLXNode>(1074)
  private val data = Array(9) { IntArray(9) }
  private lateinit var head: DLXNode

  // 数独题解个数
  val solutionCount: Int
    get() = solutions.size

  init {
    initDLX()
    initData(topic)
    solve(0)
  }

  /**
   * 初始化DLX节点
   */
  private fun initDLX() {
    val r = 730
    val c = 324
    head = DLXNode(r, c)
    head.R = head
    head.L = head.R
    head.D = head.L
    head.U = head.D
    for (i in 0 until c) {
      col[i] = DLXNode(r, i)
      col[i]!!.L = head
      col[i]!!.R = head.R
      col[i]!!.R.L = col[i]!!
      col[i]!!.L.R = col[i]!!.R.L
      col[i]!!.D = col[i]!!
      col[i]!!.U = col[i]!!.D
      size[i] = 0
    }
    for (i in r - 1 downTo 0) {
      row[i] = DLXNode(i, c)
      row[i]!!.U = head
      row[i]!!.D = head.D
      row[i]!!.D.U = row[i]!!
      row[i]!!.U.D = row[i]!!.D.U
      row[i]!!.R = row[i]!!
      row[i]!!.L = row[i]!!.R
    }
  }

  /**
   * 添加DLX节点
   *
   * @param r 行
   * @param c 列
   */
  private fun addNode(r: Int, c: Int) {
    val p = DLXNode(r, c)
    p.R = row[r]!!
    p.L = row[r]!!.L
    p.R.L = p
    p.L.R = p.R.L
    p.U = col[c]!!
    p.D = col[c]!!.D
    p.D.U = p
    p.U.D = p.D.U
    ++size[c]
  }

  private fun addNode(i: Int, j: Int, k: Int) {
    val r = (i * 9 + j) * 9 + k
    addNode(r, i * 9 + k - 1)
    addNode(r, 9 * 9 + j * 9 + k - 1)
    addNode(r, 2 * 9 * 9 + block(i, j) * 9 + k - 1)
    addNode(r, 3 * 9 * 9 + i * 9 + j)
  }

  private fun block(x: Int, y: Int) = x / 3 * 3 + y / 3

  private fun cover(c: Int) {
    if (c == THRESHOLD) {
      return
    }
    col[c]!!.delLR()
    var C = col[c]!!.D
    while (C !== col[c]) {
      if (C.c == THRESHOLD) {
        C = C.D
        continue
      }
      var R = C.L
      while (R !== C) {
        if (R.c == THRESHOLD) {
          R = R.L
          continue
        }
        --size[R.c]
        R.delUD()
        R = R.L
      }
      C.delLR()
      C = C.D
    }
  }

  private fun resume(c: Int) {
    if (c == THRESHOLD) {
      return
    }
    var C = col[c]!!.U
    while (C !== col[c]) {
      if (C.c == THRESHOLD) {
        C = C.U
        continue
      }
      C.resumeLR()
      var R = C.R
      while (R !== C) {
        if (R.c == THRESHOLD) {
          R = R.R
          continue
        }
        ++size[R.c]
        R.resumeUD()
        R = R.R
      }
      C = C.U
    }
    col[c]!!.resumeLR()
  }

  private fun solve(depth: Int): Boolean {
    if (head.L === head) {
      val solution = Array(9) { IntArray(9) }
      for (i in 0..8) {
        System.arraycopy(data[i], 0, solution[i], 0, 9)
      }
      solutions.add(solution)
      return solutions.size >= maxDepth
    }
    var minSize = 1 shl 30
    var c = -1
    var p = head.L
    while (p !== head) {
      if (size[p.c] < minSize) {
        minSize = size[p.c]
        c = p.c
      }
      p = p.L
    }
    cover(c)
    p = col[c]!!.D
    while (p !== col[c]) {
      p.R.L = p
      var cell = p.L
      while (cell !== p) {
        cover(cell.c)
        cell = cell.L
      }
      p.R.L = p.L
      val rr = p.r - 1
      data[rr / (9 * 9)][rr / 9 % 9] = rr % 9 + 1
      if (solve(depth + 1)) {
        return true
      }
      p.L.R = p
      cell = p.R
      while (cell !== p) {
        resume(cell.c)
        cell = cell.R
      }
      p.L.R = p.R
      p = p.D
    }
    resume(c)
    return false
  }

  private fun initData(topic: Array<IntArray>) {
    var i = 0
    while (i < 9) {
      var j = 0
      while (j < 9) {
        if (topic[i][j] > 0) {
          addNode(i, j, topic[i][j])
        } else {
          var k = 1
          while (k <= 9) {
            addNode(i, j, k)
            ++k
          }
        }
        ++j
      }
      ++i
    }
  }

  private class DLXNode(val r: Int, val c: Int) {
    lateinit var U: DLXNode
    lateinit var D: DLXNode
    lateinit var L: DLXNode
    lateinit var R: DLXNode
    fun delLR() {
      L.R = R
      R.L = L
    }

    fun delUD() {
      U.D = D
      D.U = U
    }

    fun resumeLR() {
      R.L = this
      L.R = R.L
    }

    fun resumeUD() {
      D.U = this
      U.D = D.U
    }
  }
}